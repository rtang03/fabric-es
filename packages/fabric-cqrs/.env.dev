CHANNEL_NAME=loanapp
CA_NAME=rca-org1
CA_ENROLLMENT_ID_ADMIN=rca-org1-admin
CA_ENROLLMENT_SECRET_ADMIN=rca-org1-adminPW
ORDERER_HOSTNAME=orderer0-org0
ORDERER_TLSCA_CERT=../../deployments/dev-net/artifacts/crypto-config/Org0MSP/orderer0.org0.com/tls-msp/tlscacerts/tls-0-0-0-0-5052.pem
CONNECTION_PROFILE=connection/connection-org1.local.yaml
MSPID=Org1MSP
ORG_ADMIN_ID=admin-org1.net
ORG_ADMIN_SECRET=admin-org1.netPW
WALLET=assets/wallet

# Org2 (pbctfp) - SOLELY used for running service.multiusers.test.ts
ORG2_ORG_ADMIN_ID=admin-org2.net
ORG2_ORG_ADMIN_SECRET=admin-org2.netPW
ORG2_MSPID=Org2MSP
ORG2_CONNECTION_PROFILE=connection/connection-org2.local.yaml
ORG1_WALLET=assets/walletOrg1
ORG2_WALLET=assets/walletOrg2
ORG2_CA_NAME=rca-org1
ORG2_CA_ENROLLMENT_ID_ADMIN=rca-org2-admin
ORG2_CA_ENROLLMENT_SECRET_ADMIN=rca-org2-adminPW



