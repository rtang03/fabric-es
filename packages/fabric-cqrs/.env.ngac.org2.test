## unsure if this is required
COLLECTION=Org1PrivateDetails
PATH_TO_CHANNEL_CONFIG=../../artifacts/channel-config

### MANDATORY
#Orgnaization Admin EnrollmentID
ORG_ADMIN_ID=Admin@org2.example.com
#Organization Admin EnrollSecret
ORG_ADMIN_SECRET=peer1pw
#URL for Orgnization CA
ORG_CA_URL=https://0.0.0.0:5055
# org name defined in connnection profile; case sensitive
ORGNAME=Org2
# Path to Fabric Network
NETWORK_LOCATION=../../network/hosts
# Path to Connection Path
CONNECTION_PROFILE=connection/peer0org2.yaml
# Relative path to wallet directory
WALLET=assets/wallet
# Channel Event Hub
CHANNEL_HUB=peer0.org2.example.com
# Optional:
CHAINCODE_ID=eventstore
CHAINCODE_ID_PRIVATEDATA=privatedata
# Used in logger
APPLICATION=peer0org2app


