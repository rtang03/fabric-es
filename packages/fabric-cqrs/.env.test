#MSPID=Org1MSP
#CHANNEL_NAME=mychannel

COLLECTION=Org1PrivateDetails
PATH_TO_CHANNEL_CONFIG=../../artifacts/channel-config
CHANNEL_NAME=eventchannel

### MANDATORY
#Orgnaization Admin EnrollmentID
ORG_ADMIN_ID=Admin@org1.example.com
#Organization Admin EnrollSecret
ORG_ADMIN_SECRET=peer1pw
#URL for Orgnization CA
ORG_CA_URL=https://0.0.0.0:5054
# org name defined in connnection profile; case sensitive
ORGNAME=Org1
# Path to Fabric Network
NETWORK_LOCATION=../../network/hosts
# Path to Connection Path
CONNECTION_PROFILE=connection/peer0org1.yaml
# Relative path to wallet directory
WALLET=assets/wallet
# Channel Event Hub
CHANNEL_HUB=peer0.org1.example.com
# Optional:
CHAINCODE_ID=eventstore
CHAINCODE_ID_PRIVATEDATA=privatedata


