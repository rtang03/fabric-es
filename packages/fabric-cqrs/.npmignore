*
!dist/
!dist/**/*
!package.json
!*.md
!LICENSE
!README.md
