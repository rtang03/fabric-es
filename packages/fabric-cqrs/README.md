# fabric-cqrs

## Installation

```shell script
yarn install @fabric-es/fabric-cqrs
```
