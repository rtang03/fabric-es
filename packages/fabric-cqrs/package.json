{
  "name": "@espresso/fabric-cqrs",
  "version": "1.0.0",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "yarn run clean && tsc",
    "build:prod": "yarn run clean && tsc -p ./tsconfig.prod.json",
    "package": "yarn run build:cleanup && yarn run cp:package.json && yarn run cp:dist",
    "build:cleanup": "rimraf ../../.build/packages/fabric-cqrs && mkdir -p ../../.build/packages/fabric-cqrs",
    "cp:package.json": "cp ./package.json ../../.build/packages/fabric-cqrs/",
    "cp:dist": "cp -R ./dist ../../.build/packages/fabric-cqrs/",
    "clean": "rimraf dist",
    "test:account": "jest account.unit.test",
    "test:service.integration": "jest service.integration.test",
    "test:service.multiusers": "rimraf assets/walletOrg1 && rimraf assets/walletOrg2 && jest service.multiusers.test",
    "test:service.privatedata.integration": "jest service.privatedata.integration.test",
    "test:service.ngac.crud.integration": "(broke) jest service.ngac.crud.integration.test",
    "test:service.ngac.pe.int.1u": "(broke) jest one-user.service.ngac.pe.int.test",
    "test:store.integration": "jest command.integration.test",
    "test:store.privatedata.integration": "jest command.privatedata.integration.test",
    "test:peer.integration": "jest peer.integration.test",
    "test:peer.privatedata.integration": "jest peer.privatedata.integration.test",
    "test:peer.ngac.integration": "(not implemented yet) jest peer.ngac.integration.test",
    "test:all-unit-test": "jest unit.test",
    "enrollAdmin": "ts-node src/account/enrollAdmin.ts",
    "enrollCaAdmin": "ts-node src/account/enrollCaAdmin.ts",
    "enrollAdminOrg2": "(broke) ts-node src/ngac_test/enrollAdmin.ts",
    "typedoc": "typedoc --excludeNotExported --theme markdown --exclude \"**/*+(.test|constants).ts\" --out docs src/index.ts"
  },
  "engines": {
    "node": ">=8",
    "npm": ">=5"
  },
  "private": true,
  "engineStrict": true,
  "dependencies": {
    "@espresso/operator": "1.0.0",
    "dotenv": "^8.1.0",
    "fabric-client": "^1.4.4",
    "fabric-contract-api": "^1.4.2",
    "fabric-network": "^1.4.4",
    "js-yaml": "^3.13.1",
    "lodash": "^4.17.15",
    "prettyjson": "^1.2.1",
    "redux": "^4.0.4",
    "redux-observable": "^1.2.0",
    "rxjs": "^6.5.3",
    "winston": "^3.2.1"
  },
  "devDependencies": {
    "@types/bytebuffer": "^5.0.40",
    "@types/dotenv": "^6.1.1",
    "@types/jest": "^24.0.18",
    "@types/js-yaml": "^3.12.1",
    "@types/lodash": "^4.14.138",
    "@types/node": "^8.10.54",
    "jest": "^24.9.0",
    "prettier": "^1.19.1",
    "rimraf": "^3.0.0",
    "ts-jest": "^24.0.2",
    "ts-node": "^8.3.0",
    "tslint": "^5.20.0",
    "tslint-config-prettier": "^1.18.0",
    "tslint-plugin-prettier": "^2.0.1",
    "typedoc": "^0.15.7",
    "typedoc-plugin-markdown": "^2.2.16",
    "typescript": "^3.7.2"
  }
}
