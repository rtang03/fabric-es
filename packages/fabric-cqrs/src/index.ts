/**
 * @packageDocumentation
 * @hidden
 */
export {
  createPeer,
  getMockRepository,
  getPrivatedataMockRepository,
  createProjectionDb,
  createQueryDatabase
} from './peer';
export { getContract, getNetwork, submit, evaluate, evaluate$ } from './services';
export * from './types';
export {
  dispatchResult,
  generateToken,
  getAction,
  getErrorAction,
  getErrorActionHandler,
  getSuccessAction,
  getSuccessActionHandler
} from './cqrs/utils';
export { fromCommitsToGroupByEntityId, getHistory, isCommit } from './peer/utils';
