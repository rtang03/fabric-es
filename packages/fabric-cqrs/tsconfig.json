{
  "compilerOptions": {
    "outDir": "dist",
    "target": "es2017",
    "moduleResolution": "node",
    "esModuleInterop": true,
    "module": "commonjs",
    "declaration": true,
    "sourceMap": false,
    "types": ["node", "jest"],
    "lib": ["es2017", "dom", "esnext.asynciterable"],
    "rootDir": "src",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true
  },
  "exclude": ["node_modules", "dist"],
  "typedocOptions": {
    "entryPoints": ["src/index.ts"],
    "out": "docs",
    "name": "fabric-cqrs",
    "includeVersion": true,
    "excludeExternals": true,
    "exclude": "\"**/*+(.test|constants|index).ts\"",
    "readme": "../../README.md",
    "hideGenerator": true
  }
}
