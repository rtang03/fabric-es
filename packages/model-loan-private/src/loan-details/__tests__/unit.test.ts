import { buildFederatedSchema } from '@apollo/federation';
import { Commit, getPrivatedataMockRepository, getReducer, PrivatedataRepository } from '@espresso/fabric-cqrs';
import { DataSrc } from '@espresso/gw-node';
// import { APPLY_LOAN, Loan, loanResolvers, loanTypeDefs, LoanEvents, loanReducer } from '@espresso/model-loan';
import { ApolloServer } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';
import {
  CREATE_LOAN_DETAILS, GET_DETAILS_BY_ID,
  LoanDetails, LoanDetailsEvents, loanDetailsReducer, loanDetailsResolvers, loanDetailsTypeDefs,
  UPDATE_LOAN_DETAILS
} from '..';

const userId = 'unitTestUser';
// const loanId = 'L1234';
const mockdb: Record<string, Commit> = {};
// const loanRepo = getMockRepository<Loan, LoanEvents>(mockdb, 'loan', getReducer<Loan, LoanEvents>(loanReducer));
const loanDetailsRepo: PrivatedataRepository = getPrivatedataMockRepository<LoanDetails, LoanDetailsEvents>(
  mockdb, 'loanDetails', getReducer<LoanDetails, LoanDetailsEvents>(loanDetailsReducer)
);

let service;

beforeAll(async () => {
  // const server = new ApolloServer({
  //   schema: buildFederatedSchema([{ typeDefs: loanTypeDefs, resolvers: loanResolvers }]),
  //   dataSources: () => ({
  //     loan: new DataSrc({ repo: loanRepo })
  //   }),
  //   context: () => ({ enrollmentId: 'admin' })
  // });

  // await createTestClient(server).mutate({
  //   mutation: APPLY_LOAN,
  //   variables: {
  //     userId, loanId,
  //     description: 'Unit test loan 1234',
  //     reference: 'REF-UNIT-TEST-LOAN-X',
  //     comment: 'Hello 1234'
  //   }});

  service = new ApolloServer({
    schema: buildFederatedSchema([{ typeDefs: loanDetailsTypeDefs, resolvers: loanDetailsResolvers }]),
    dataSources: () => ({
      loanDetails: new DataSrc({ repo: loanDetailsRepo })
    }),
    context: () => ({ enrollmentId: 'admin' })
  });
});

afterAll(async () =>
  new Promise(done =>
    setTimeout(() => {
      console.log('Loan Unit Test - Resolver Finished');
      done();
    }, 500)
));

describe('LoanDetails Unit Test - Resolver', () => {
  it('create loanDetails 1', async () =>
    createTestClient(service).mutate({
      mutation: CREATE_LOAN_DETAILS,
      variables: {
        userId, loanId: 'L0001',
        registration: 'REG-0001',
        companyName: 'Loan Requester 1',
        contactName: 'Contact 1',
        contactPhone: '555-0001',
        contactEmail: 'c0001@fake.it',
        loanType: 'Post-Shipment',
        startDate: '1542385275431',
        tenor: 76,
        currency: 'HKD',
        requestedAmt: 50000
      }})
    .then(({ data }) => expect(data.createLoanDetails.id).toEqual('L0001'))
    .catch(_ => expect(false).toBeTruthy())
  );

  it('create loanDetails 2', async () =>
    createTestClient(service).mutate({
      mutation: CREATE_LOAN_DETAILS,
      variables: {
        userId, loanId: 'L0002', registration: 'REG-0002', companyName: 'Loan Requester 2',
        contactName: 'Contact 2', contactPhone: '555-0002', contactEmail: 'c0002@fake.it',
        loanType: 'Post-Shipment', startDate: '1542385275432',
        tenor: 76, currency: 'WON', requestedAmt: 50000, comment: 'Yello 0002'
      }})
    .then(({ data }) => expect(data.createLoanDetails.id).toEqual('L0002'))
    .catch(_ => expect(false).toBeTruthy())
  );

  it('create loanDetails 3', async () =>
    createTestClient(service).mutate({
      mutation: CREATE_LOAN_DETAILS,
      variables: {
        userId, loanId: 'L0003', registration: 'REG-0003', companyName: 'Loan Requester 3',
        contactName: 'Contact 3', contactPhone: '555-0003', contactEmail: 'c0003@fake.it',
        loanType: 'Post-Shipment', startDate: '1542385275433',
        tenor: 76, currency: 'HKD', requestedAmt: 50000, comment: 'Yello 0003'
      }})
    .then(({ data }) => expect(data.createLoanDetails.id).toEqual('L0003'))
    .catch(_ => expect(false).toBeTruthy())
  );

  it('update loanDetails 2', async () =>
    createTestClient(service).mutate({
      mutation: UPDATE_LOAN_DETAILS,
      variables: {
        userId, loanId: 'L0002',
        currency: 'HKD', comment: 'Hello 0002'
      }})
    .then(({ data }) => expect(data.updateLoanDetails.map(d => (d && d.id) ? d.id : '')).toContain('L0002'))
    .catch(_ => expect(false).toBeTruthy())
  );

  it('update an non-existing loanDetails', async () =>
    createTestClient(service).mutate({
      mutation: UPDATE_LOAN_DETAILS,
      variables: {
        userId, loanId: 'L9999', comment: 'Hello 9999'
      }})
    .then(({ errors }) => expect(errors.reduce((acc, cur) =>
      cur.message.includes('LOAN_DETAILS_NOT_FOUND') ? cur.message : acc, '')).toContain('LOAN_DETAILS_NOT_FOUND'))
    .catch(_ => expect(false).toBeTruthy())
  );

  it ('update loanDetails 1 with empty currency', async () =>
    createTestClient(service).mutate({
      mutation: UPDATE_LOAN_DETAILS,
      variables: {
        userId, loanId: 'L0001', currency: ''
      }})
    .then(({ errors }) => expect(errors.reduce((acc, cur) =>
      cur.message.includes('REQUIRED_DATA_MISSING') ? cur.message : acc, '')).toContain('REQUIRED_DATA_MISSING'))
    .catch(_ => expect(false).toBeTruthy())
  );

  it('update loanDetails 3 with both successful and fail cases', async () =>
    createTestClient(service).mutate({
      mutation: UPDATE_LOAN_DETAILS,
      variables: {
        userId, loanId: 'L0003', startDate: '1542385275430', currency: '', comment: 'Hello 0003'
      }})
    .then(data => {
      console.log('MOMO', JSON.stringify(data));
      return data;
    })
    .then(({ data, errors }) => {
      const errs = errors.map(e => e.message);
      expect(errs).toContain('Error: INVALID_OPERATION');
      expect(errs).toContain('Error: REQUIRED_DATA_MISSING');
      expect(data.updateLoanDetails.map(d => (d && d.id) ? d.id : '')).toContain('L0003');
    })
    .catch(_ => expect(false).toBeTruthy())
  );

  it('query loanDetails 1 by id', async () =>
    createTestClient(service).query({
      query: GET_DETAILS_BY_ID,
      variables: { loanId: 'L0001' }
    })
    .then(({ data }) => expect(data.getLoanDetailsById).toMatchSnapshot())
    .catch(_ => expect(false).toBeTruthy())
  );

  it('query loanDetails 2 by id', async () =>
    createTestClient(service).query({
      query: GET_DETAILS_BY_ID,
      variables: { loanId: 'L0002' }
    })
    .then(({ data }) => expect(data.getLoanDetailsById).toMatchSnapshot())
    .catch(_ => expect(false).toBeTruthy())
  );

  it('query loanDetails 3 by id', async () =>
    createTestClient(service).query({
      query: GET_DETAILS_BY_ID,
      variables: { loanId: 'L0003' }
    })
    .then(({ data }) => expect(data.getLoanDetailsById).toMatchSnapshot())
    .catch(_ => expect(false).toBeTruthy())
  );
});