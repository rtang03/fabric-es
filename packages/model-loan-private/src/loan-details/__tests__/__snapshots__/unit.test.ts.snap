// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanDetails Unit Test - Resolver query loanDetails 1 by id 1`] = `
Object {
  "approvedAmt": null,
  "comment": null,
  "contact": Object {
    "email": "c0001@fake.it",
    "name": "Contact 1",
    "phone": "555-0001",
    "salutation": null,
    "title": null,
  },
  "currency": "HKD",
  "loanId": "L0001",
  "loanType": "Post-Shipment",
  "requestedAmt": 50000,
  "requester": Object {
    "name": "Loan Requester 1",
    "registration": "REG-0001",
    "type": null,
  },
  "startDate": "1542385275431",
  "tenor": 76,
}
`;

exports[`LoanDetails Unit Test - Resolver query loanDetails 2 by id 1`] = `
Object {
  "approvedAmt": null,
  "comment": "Hello 0002",
  "contact": Object {
    "email": "c0002@fake.it",
    "name": "Contact 2",
    "phone": "555-0002",
    "salutation": null,
    "title": null,
  },
  "currency": "HKD",
  "loanId": "L0002",
  "loanType": "Post-Shipment",
  "requestedAmt": 50000,
  "requester": Object {
    "name": "Loan Requester 2",
    "registration": "REG-0002",
    "type": null,
  },
  "startDate": "1542385275432",
  "tenor": 76,
}
`;

exports[`LoanDetails Unit Test - Resolver query loanDetails 3 by id 1`] = `
Object {
  "approvedAmt": null,
  "comment": "Hello 0003",
  "contact": Object {
    "email": "c0003@fake.it",
    "name": "Contact 3",
    "phone": "555-0003",
    "salutation": null,
    "title": null,
  },
  "currency": "HKD",
  "loanId": "L0003",
  "loanType": "Post-Shipment",
  "requestedAmt": 50000,
  "requester": Object {
    "name": "Loan Requester 3",
    "registration": "REG-0003",
    "type": null,
  },
  "startDate": "1542385275433",
  "tenor": 76,
}
`;
